Sources/SwiftSyntax/SyntaxParser.swift

```diff
@@ -0,0 +1,224 @@
+    assert(source.isNativeUTF8)
+
+    let c_parser = swiftparse_parser_create()
+    defer {
+      swiftparse_parser_dispose(c_parser)
+    }
+
+    let nodeHandler = { (cnode: CSyntaxNodePtr!) -> UnsafeMutableRawPointer in
+      let node = makeRawNode(cnode, source: source)
+      let bits = Unmanaged.passRetained(node)
+      return bits.toOpaque()
+    }
+    swiftparse_parser_set_node_handler(c_parser, nodeHandler);
+
+    if let parseLookup = parseLookup {
+      let nodeLookup = {
+            (offset: Int, kind: CSyntaxKind) -> CParseLookupResult in
+        guard let foundNode =
+            parseLookup.lookUp(offset,
+                               kind: SyntaxKind.fromRawValue(kind)) else {
+          return CParseLookupResult(length: 0, node: nil)
+        }
+        let lengthToSkip = foundNode.byteSize
+        let bits = Unmanaged.passRetained(foundNode.raw)
+        return CParseLookupResult(length: lengthToSkip, node: bits.toOpaque())
+      }
+      swiftparse_parser_set_node_lookup(c_parser, nodeLookup);
+    }
+
+    let c_top = swiftparse_parse_string(c_parser, source)
+    return moveFromCRawNode(c_top)!
+  }
+}
+
+fileprivate
+func makeRawToken(_ c_node: CSyntaxNode, source: String) -> RawSyntax {
+  let tokdat = c_node.token_data
+  let kind = tokdat.kind
+  var leadingTriviaLen = 0
+  for i in 0..<Int(tokdat.leading_trivia_count) {
+    leadingTriviaLen += Int(tokdat.leading_trivia![i].length)
+  }
+  var trailingTriviaLen = 0
+  for i in 0..<Int(tokdat.trailing_trivia_count) {
+    trailingTriviaLen += Int(tokdat.trailing_trivia![i].length)
+  }
+
+  let offset = Int(c_node.range.offset)
+  let tokOffset = offset + leadingTriviaLen
+  let tokLen = Int(c_node.range.length) - leadingTriviaLen - trailingTriviaLen
+
+  let text = source.utf8Slice(offset: tokOffset, length: tokLen)
+  let tokKind = try! TokenKind.fromRawValue(kind: kind, text: text)
+  let leadingTrivia = toTrivia(tokdat.leading_trivia,
+                               count: Int(tokdat.leading_trivia_count),
+                               offset: offset, source: source)
+  let trailingTrivia = toTrivia(tokdat.trailing_trivia,
+                                count: Int(tokdat.trailing_trivia_count),
+                                offset: tokOffset+tokLen, source: source)
+  let presence: SourcePresence = c_node.present ? .present : .missing
+  return RawSyntax(kind: tokKind, leadingTrivia: leadingTrivia,
+                   trailingTrivia: trailingTrivia, presence: presence)
+}
+
+fileprivate
+func makeRawNode(_ cnodeptr: CSyntaxNodePtr, source: String) -> RawSyntax {
+  let cnode = cnodeptr.pointee
+  let kind = SyntaxKind.fromRawValue(cnode.kind)
+  if kind == .token {
+    return makeRawToken(cnode, source: source)
+  } else {
+    var layout = [RawSyntax?]()
+    layout.reserveCapacity(Int(cnode.layout_data.nodes_count))
+    for i in 0..<Int(cnode.layout_data.nodes_count) {
+      let subnode = moveFromCRawNode(cnode.layout_data.nodes![i])
```

Comment:
We are able to move these nodes from C because nodes are feeding in bottom-up fashion. Could you add some comments about this assumption?

Reply:
Done.

Reply:
Thank you!